### ISSUE TITLE
[db] Deadlock in the db upon repeated execution of remove queries

### ISSUE THREAD

**Issue by pinghe** [agnesoft/agdb#1130]
I've encountered a performance issue that I believe needs to be addressed. When executing queries from three JSON files in sequence, the process works as expected initially. However, upon repeated execution of this sequence, the agdb process experiences sustained high CPU usage and does not return to normal levels.

Description of the Issue:
The issue arises when the following sequence of commands is executed repeatedly:

test_data0.json
test_data1.json
test_data2.json
Initially, the execution is successful and the system behaves as expected. However, after several iterations, the CPU usage by the agdb process remains consistently high and does not return to normal.

Steps to Reproduce:

Execute the queries from test_data0.json.
Follow with executing the queries from test_data1.json.
Then execute the queries from test_data2.json.
Repeat the sequence of steps 1-3 multiple times.
Expected Result:
The agdb process should maintain normal CPU usage levels even after repeated execution of the query sequence.

Actual Result:
After several repetitions, the CPU usage by the agdb process remains high and does not return to normal, indicating a potential performance bottleneck or resource leak.

I have monitored the CPU usage using and observed the consistent high usage.
No additional error messages or warnings are produced during this time.
I would appreciate it if you could look into this issue and provide guidance on how to resolve it or if there are any known workarounds.

Thank you for your attention to this matter.

Best regards,
pinghe

[test_data0.json](https://github.com/user-attachments/files/15901487/test_data0.json)
[test_data1.json](https://github.com/user-attachments/files/15901488/test_data1.json)
[test_data2.json](https://github.com/user-attachments/files/15901490/test_data2.json)

**Comment by pinghe** [comment]
![image](https://github.com/agnesoft/agdb/assets/377955/d34544cb-86a3-4126-a32c-0c0326035004)

![image](https://github.com/agnesoft/agdb/assets/377955/8e9e8f4e-86a3-46b2-8e97-099ea47a6482)

**Comment by michaelvlach** [comment]
I can reproduce this. From the initial investigation it looks like a deadlock in the server processing. At some point the queries simply stop being processed and the ongoing fight for some lock is causing the CPU spike. I will look into it more and get it fixed.

**Comment by michaelvlach** [comment]
@pinghe  This is now fixed. It was a tough one as it turned out to be 2 bugs in the very low levels of the database collections (multimap implementation). The issue was that in case of repeated insertions & deletions the entire map could run out slots marked "empty". Such slots were used to stop iterating over the map but because the map contained none of them it was stuck in an infinite loop. The second bug was that the "deleted" entries were not considered for insertions. Both fixed now and I have verified the provided queries work fine even with many iterations.
